app/tests/heartbeat_storage_desktop_test.cc

```diff
@@ -0,0 +1,210 @@
+  bool write_ok = storage1.Write(heartbeats);
+  ASSERT_TRUE(write_ok) << "Unable to write the heartbeat file:\n  " +
+                               storage1.GetError();
+
+  HeartbeatStorageDesktop storage2 = HeartbeatStorageDesktop(app_id);
+  LoggedHeartbeats read_heartbeats;
+  bool read_ok = storage2.ReadTo(&read_heartbeats);
+  ASSERT_TRUE(read_ok) << "Unable to read the heartbeat file:\n  " +
+                              storage2.GetError();
+  EXPECT_EQ(read_heartbeats.last_logged_date, date2);
+  ASSERT_EQ(read_heartbeats.heartbeats[user_agent].size(), 2);
+  EXPECT_EQ(read_heartbeats.heartbeats[user_agent][0], date1);
+  EXPECT_EQ(read_heartbeats.heartbeats[user_agent][1], date2);
+}
+
+TEST_F(HeartbeatStorageDesktopTest, WriteAndReadDifferentAppIds) {
+  std::string app_id = "app_id";
+  std::string different_app_id = "different_app_id";
+  std::string user_agent1 = "user_agent1";
+  std::string user_agent2 = "user_agent2";
+  std::string date1 = "2022-01-01";
+  std::string date2 = "2022-02-02";
+  // Write using app_id
+  HeartbeatStorageDesktop storage1 = HeartbeatStorageDesktop(app_id);
+  LoggedHeartbeats heartbeats1;
+  heartbeats1.last_logged_date = date1;
+  heartbeats1.heartbeats[user_agent1].push_back(date1);
+  bool write_ok = storage1.Write(heartbeats1);
+  ASSERT_TRUE(write_ok) << "Unable to write the heartbeat file:\n  " +
+                               storage1.GetError();
+
+  // Write different heartbeats using different_app_id
+  HeartbeatStorageDesktop storage2 = HeartbeatStorageDesktop(different_app_id);
+  LoggedHeartbeats heartbeats2;
+  heartbeats2.last_logged_date = date2;
+  heartbeats2.heartbeats[user_agent2].push_back(date2);
+  write_ok = storage2.Write(heartbeats2);
+  ASSERT_TRUE(write_ok) << "Unable to write the heartbeat file:\n  " +
+                               storage2.GetError();
+
+  // Read using app_id and verify it contains the original heartbeats.
+  LoggedHeartbeats read_heartbeats;
+  bool read_ok = storage1.ReadTo(&read_heartbeats);
+  ASSERT_TRUE(read_ok) << "Unable to read the heartbeat file:\n  " +
+                              storage1.GetError();
+  EXPECT_EQ(read_heartbeats.last_logged_date, date1);
+  ASSERT_EQ(read_heartbeats.heartbeats[user_agent1].size(), 1);
+  EXPECT_EQ(read_heartbeats.heartbeats[user_agent1][0], date1);
+}
+
+TEST_F(HeartbeatStorageDesktopTest, ReadNonexistentFile) {
+  // ReadHeartbeats should return a default instance
+  HeartbeatStorageDesktop storage =
+      HeartbeatStorageDesktop("nonexistent_app_id");
+  LoggedHeartbeats read_heartbeats;
+  bool read_ok = storage.ReadTo(&read_heartbeats);
+  ASSERT_TRUE(read_ok) << "Unable to read the heartbeat file:\n  " +
+                              storage.GetError();
+  EXPECT_EQ(read_heartbeats.last_logged_date, "");
+  ASSERT_EQ(read_heartbeats.heartbeats.size(), 0);
+}
+
+TEST_F(HeartbeatStorageDesktopTest, WriteAndReadWithSymbolsInAppId) {
+  std::string app_id = "start/\\?%*:|\"<>.,;=end";
```

Comment:
Rather than exposing a static method that constructs a filename, I just provided a getter to the filename that is used by an instance of storage. I also updated the test to check the generated filename when the appid has symbols.